File: 3. Longest Substring Without Repeating Characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = 0
        seen = {}
        max_length = 0

        for right in range(len(s)):
            char = s[right]
            if char in seen:  # Found duplicate
                left = max(left, seen[char] + 1)  # Move left to the furthest point where a duplicate is currently known
            seen[char] = right  # always update current info
            max_length = max(max_length, right - left + 1)

        return max_length

File: 3. Longest Substring Without Repeating Characters/solution.c
#define MAX(a,b) ((a) > (b)? (a) : (b))

int lengthOfLongestSubstring(char* s) {
    int left = 0;
    int seen[128] = {-1};  // seen[ascii_value] = index_seen
    memset(seen, -1, sizeof(seen));
    int max_length = 0;

    int len = strlen(s);
    for (int right = 0; right < len; right++) {
        char c = s[right];
        // Check if we've seen this before and it's within the window
        if (seen[c] != -1 && seen[c] >= left) {
            // If yes, update the left pointer
            left = seen[c] + 1;  // NOTE1
        }
        // Update seen array
        seen[c] = right;
        max_length = MAX(max_length, right - left + 1);
    }

    return max_length;
}

/*
NOTE1: In C, characters are internally stored as an integer, so they already "are" the ASCII value.
       The integer values can be retrieved implicitly (as is the case in seen[c]) or by explicit casting.
*/
